webapp/src/app.tsx

```diff
@@ -1936,6 +1918,166 @@ function initLogin() {
+            agreeLbl: lf("Update"),
+            disagreeLbl: lf("Quit app"),
+            disagreeClass: "red",
+            size: "medium"
+        }).then(b => {
+            if (!b) {
+                pxt.tickEvent("update.refusedCritical");
+                sendElectronMessage("quit");
+            } else {
+                pxt.tickEvent("update.acceptedCritical");
+                showDownloading();
+                sendElectronMessage("update", {
+                    targetVersion: args.targetVersion,
+                    isCritical: true
+                });
+            }
+        });
+    }
+
+    function onUpdateAvailable(args: any) {
+        let header = lf("Version {0} available", args.targetVersion);
+
+        if (args.isBeta) {
+            header += " " + lf("(beta release)");
+        }
+
+        core.confirmAsync({
+            header,
+            body: lf("A new version of {0} is ready to download and install. The app will restart during the update. Update now?", args.appName || lf("this app")),
+            agreeLbl: lf("Update"),
+            disagreeLbl: lf("Not now"),
+            size: "medium"
+        }).then(b => {
+            if (!b) {
+                if (args.isInitialCheck) {
+                    pxt.tickEvent("update.refusedInitial");
+                } else {
+                    pxt.tickEvent("update.refused");
+                }
+            } else {
+                pxt.tickEvent("update.accepted");
+                showDownloading();
+                sendElectronMessage("update", {
+                    targetVersion: args.targetVersion
+                });
+            }
+        });
+    }
+
+    function onUpdateNotAvailable() {
+        core.confirmAsync({
+            body: lf("You are using the latest version available."),
+            header: lf("Good to go!"),
+            agreeLbl: lf("Ok"),
+            hideCancel: true
+        });
+    }
+
+    function onUpdateCheckError() {
+        displayUpdateError(lf("Unable to check for update"), lf("Ok"));
+    }
+
+    function onUpdateDownloadError(args: any) {
+        let isCritical = args && args.isCritical;
+
+        hideDownloading();
+        displayUpdateError(lf("There was an error downloading the update"), isCritical ? lf("Quit app") : lf("Ok"))
+            .finally(() => {
+                if (isCritical) {
+                    sendElectronMessage("quit");
+                }
+            });
+    }
+
+    function displayUpdateError(header: string, btnLabel: string) {
+        return core.confirmAsync({
+            header,
+            body: lf("Please ensure you are connected to the Internet and try again later."),
+            agreeClass: "red",
+            agreeIcon: "cancel",
+            agreeLbl: btnLabel,
+            hideCancel: true
+        });
+    }
+
+    pxt.log('initializing electron socket');
+    electronSocket = new WebSocket('ws://localhost:3233/' + Cloud.localToken + '/electron');
```

Comment:
Good point, I just need to move the `initSocketServer()` call outside of the `initSerialMonitor()` function in server.ts